jib-core/src/main/java/com/google/cloud/tools/jib/image/ReproducibleLayerBuilder.java

```diff
@@ -34,31 +40,75 @@
+    private final Set<String> names = new HashSet<>();
+
+    /**
+     * Adds a {@link TarArchiveEntry} if its extraction path does not exist yet. Also adds all of
+     * the parent directories on the extraction path.
+     *
+     * @param tarArchiveEntry the {@link TarArchiveEntry}
+     */
+    private void add(TarArchiveEntry tarArchiveEntry) {
+      if (names.contains(tarArchiveEntry.getName())) {
+        return;
+      }
+
+      // Adds all directories along extraction paths to explicitly set permissions for those
+      // directories.
+      Path namePath = Paths.get(tarArchiveEntry.getName());
+      if (namePath.getParent() != namePath.getRoot()) {
+        add(new TarArchiveEntry(DIRECTORY_FILE, namePath.getParent().toString()));
+      }
+
+      entries.add(tarArchiveEntry);
+      names.add(tarArchiveEntry.getName());
+    }
+
+    private List<TarArchiveEntry> getSortedEntries() {
+      List<TarArchiveEntry> sortedEntries = new ArrayList<>(entries);
+      sortedEntries.sort(Comparator.comparing(TarArchiveEntry::getName));
+      return sortedEntries;
+    }
+  }
+
   /**
    * Builds the {@link TarArchiveEntry}s for adding this {@link LayerEntry} to a tarball archive.
    *
    * @return the list of {@link TarArchiveEntry}
    * @throws IOException if walking a source file that is a directory failed
    */
-  private static List<TarArchiveEntry> buildAsTarArchiveEntries(LayerEntry layerEntry)
-      throws IOException {
+  @VisibleForTesting
+  static List<TarArchiveEntry> buildAsTarArchiveEntries(LayerEntry layerEntry) throws IOException {
     List<TarArchiveEntry> tarArchiveEntries = new ArrayList<>();
 
+    // Adds the files to extract relative to the extraction path.
     for (Path sourceFile : layerEntry.getSourceFiles()) {
       if (Files.isDirectory(sourceFile)) {
         new DirectoryWalker(sourceFile)
             .filterRoot()
+            .filter(path -> !Files.isDirectory(path))
             .walk(
                 path -> {
-                  /*
-                   * Builds the same file path as in the source file for extraction. The iteration
-                   * is necessary because the path needs to be in Unix-style.
-                   */
-                  StringBuilder subExtractionPath =
-                      new StringBuilder(layerEntry.getExtractionPath());
+                  // Builds the same file path as in the source file for extraction. The iteration
+                  // is necessary because the path needs to be in Unix-style.
+                  Path subExtractionPath = Paths.get(layerEntry.getExtractionPath());
                   Path sourceFileRelativePath = sourceFile.getParent().relativize(path);
                   for (Path sourceFileRelativePathComponent : sourceFileRelativePath) {
-                    subExtractionPath.append('/').append(sourceFileRelativePathComponent);
+                    subExtractionPath = subExtractionPath.resolve(sourceFileRelativePathComponent);
```

Comment:
Wasn't the point of using `append('/')` instead of `resolve` for forcing the paths to be Unix-style? (Comment on line 107)

Reply:
Actually, that seems to be a relic of the old way - removing.